### PR TITLE
Rsar support with small fixes from khang06 

### PR DESCRIPTION
**Pull request by MDHEXT**
This adds support for RSAR files and has some small tweaks made by khang06. BRSAR files are the SDAT files of the Wii, they contain sequences and sfx information. This pull request adds support for RSAR files in vgmtrans and allows midi files and soundbanks to be dumped using vgmtrans. This wouldn't have been possible without kode54. He takes all the credit for making the master branch and magicus's changes work.

### CODE REVIEW

**Comment by MDHEXT** [comment]
The build succeeded in my repository... I'm not sure why it failed here.
Appveyor: https://ci.appveyor.com/project/MDHEXT/vgmtrans/builds/30993237/artifacts

**Comment by loveemu** [comment]
Thank you for your contribution.

@sykhro Could you handle this if possible?

**Comment by sykhro** [comment]
I'm out for a conference in these couple days so it might take a bit. I'll start review soon, overall I think it has the same concerns as the other PR.
As for CI, it's something to fix in the appveyor file.

**Comment by MDHEXT** [comment]
it seems like it can't open revokart.brsar (Mario Kart Wii). Can you guys confirm it doesn't work for you guys either? I know that magicus's original fork could open it sometimes. I have two downloads of it and only the one in my downloads folder works.

File: src/main/DLSFile.cpp [hunk 1]
```diff
@@ -463,14 +463,39 @@ uint32_t DLSWave::GetSize() {
...
+
+  uint8_t bytesPerSample = wBitsPerSample / 8;
+  uint8_t skip = wChannels * bytesPerSample;
+  for (uint32_t i = whichChannel * bytesPerSample; i < dataSize; i += skip) {
+    buf.insert(buf.end(), &data[i], &data[i + bytesPerSample]);
+  }
+}
+
+void DLSWave::WriteSample(vector<uint8_t> &buf, bool mono) {
+  if (false && (wChannels != 1 && mono)) {
```
Comment (sykhro) [inline src/main/DLSFile.cpp]: Why is this stubbed like this? DLS (at least the current version we're using) mandates mono samples. We want to be spec-compliant to avoid DAWs and whatnot refusing our files.

File: src/main/DLSFile.cpp [hunk 2]
```diff
@@ -463,14 +463,39 @@ uint32_t DLSWave::GetSize() {
...
+  } else {
+    PushTypeOnVect<uint32_t>(buf, dataSize);                //size: this is the ACTUAL size, not the even-aligned size
+    if (dataSize % 2)
+      buf.push_back(0);
+    buf.insert(buf.end(), data, data + dataSize);           //Write the sample
+  }
+}
+
+void DLSWave::Write(vector<uint8_t> &buf, bool mono) {
+  //mono = false;
```
Comment (sykhro) [inline src/main/DLSFile.cpp]: ditto

File: src/main/DLSFile.h [hunk 3]
```diff
@@ -90,7 +90,7 @@ class DLSFile: public RiffFile {
 
   uint32_t GetSize(void);
 
-  int WriteDLSToBuffer(std::vector<uint8_t> &buf);
+  int WriteDLSToBuffer(std::vector<uint8_t> &buf, bool mono = false);
```
Comment (sykhro) [inline src/main/DLSFile.h]: ditto again

File: src/main/VGMColl.cpp [hunk 4]
```diff
@@ -137,9 +137,9 @@ void VGMColl::UnpackSampColl(SynthFile &synthfile, VGMSampColl *sampColl, vector
       sampInfo->SetLoopInfo(samp->loop, samp);
 
     double attenuation = (samp->volume != -1) ? ConvertLogScaleValToAtten(samp->volume) : 0;
-    uint8_t unityKey = (samp->unityKey != -1) ? samp->unityKey : 0x3C;
+    // uint8_t unityKey = (samp->unityKey != -1) ? samp->unityKey : 0x3C;
     short fineTune = samp->fineTune;
-    sampInfo->SetPitchInfo(unityKey, fineTune, attenuation);
+    sampInfo->SetPitchInfo(0x3C, fineTune, attenuation);
```
Comment (sykhro) [inline src/main/VGMColl.cpp]: What's the point of this? It seems broken.

File: src/main/VGMColl.cpp [hunk 5]
```diff
@@ -593,3 +593,32 @@ bool VGMColl::OnSaveAllSF2() {
...
+    }
+    else
+      pRoot->AddLogItem(new LogItem(std::wstring(L"Failed to save DLS file"), LOG_LEVEL_ERR, L"VGMColl"));
+
+    filepath = dirpath + L"\\" + ConvertToSafeFileName(this->name) + L".mid";
+    if (!this->seq->SaveAsMidi(filepath))
+      pRoot->AddLogItem(new LogItem(std::wstring(L"Failed to save MIDI file"), LOG_LEVEL_ERR, L"VGMColl"));
+  }
+  return true;
+}
```
Comment (sykhro) [inline src/main/VGMColl.cpp]: useful, doesn't belong in this PR though

File: src/main/VGMColl.h [hunk 6]
```diff
@@ -17,6 +17,7 @@ class VGMColl
   BEGIN_MENU(VGMColl)
       MENU_ITEM(VGMColl, OnSaveAllDLS, L"Save as MIDI and DLS.")
       MENU_ITEM(VGMColl, OnSaveAllSF2, L"Save as MIDI and SoundFont 2.")
+      MENU_ITEM(VGMColl, OnSaveAllBoth, L"Save as MIDI, DLS, SF2.")
```
Comment (sykhro) [inline src/main/VGMColl.h]: ditto

File: src/main/formats/NDSInstrSet.cpp [hunk 7]
```diff
@@ -70,13 +70,19 @@ bool NDSInstr::LoadInstr() {
...
       };
-      name = L"NES Sq (" + dutyCycles[dutyCycle] + L")";
+      name = L"PSG Sq (" + dutyCycles[dutyCycle] + L")";
       unLength = 10;
+      VGMRgn *rgn = AddRgn(dwOffset, 10, dutyCycle);
+      /* XXX: This is disgusting. Rely on the PSG sample coll being the first one. */
+      rgn->sampCollPtr = nullptr;
+      GetArticData(rgn, dwOffset + 4);
+      /* A3 */
+      rgn->SetUnityKey(45);
```
Comment (sykhro) [inline src/main/formats/NDSInstrSet.cpp]: This needs a less-hacky alternative.

File: src/main/formats/NDSScanner.cpp [hunk 8]
```diff
@@ -27,6 +27,85 @@ void NDSScanner::SearchForSDAT(RawFile *file) {
...
+  PSGSampColl(RawFile *file) : VGMSampColl(NDSFormat::name, file, 0, 0, L"PSG Samples") {}
+private:
+  virtual bool GetSampleInfo() {
+    for (uint8_t i = 0; i < 8; i++)
+      samples.push_back(new PSGDutySamp(this, i));
+    return true;
+  }
+};
+
+
```
Comment (sykhro) [inline src/main/formats/NDSScanner.cpp]: Not in this PR please. Also, why is it part of just NDSScanner? It could be useful for other things.

File: src/ui/windows/HexView.cpp [hunk 9]
```diff
@@ -751,7 +751,7 @@ void CHexView::SetupColor()
 	{
 		//if (offset+j > curFile->dwOffset+curFile->unLength)
 		//	break;
-		VGMItem* item = curFile->GetItemFromOffset(offset+j, false);
+          VGMItem* item = nullptr; // curFile->GetItemFromOffset(offset + j, false);
```
Comment (sykhro) [inline src/ui/windows/HexView.cpp]: We don't nuke something like this.

**Comment by MDHEXT** [comment]
I can see how hacky it is, but I didn't code it. I just got khang06's fork of magicus' rsar vgmtrans edit because it says it had more compatibility with rsar seqs. I then made sure it could merge cleanly with the master branch without many issues. I basically only made it mergable.

**Comment by MDHEXT** [comment]
I can see how hacky it is, but I didn't code it. I just got khang06's fork of magicus' rsar vgmtrans edit because it says it had more compatibility with rsar seqs. I then made sure it could merge cleanly with the master branch without many issues. I basically only made it mergable.

**Comment by MDHEXT** [comment]
I can see how hacky it is, but I didn't code it. I just got khang06's fork of magicus' rsar vgmtrans edit because it says it had more compatibility with rsar seqs. I then made sure it could merge cleanly with the master branch without many issues. I basically only made it mergable.

**Comment by MDHEXT** [comment]
I can see how hacky it is, but I didn't code it. I just got khang06's fork of magicus' rsar vgmtrans edit because it says it had more compatibility with rsar seqs. I then made sure it could merge cleanly with the master branch without many issues. I basically only made it mergable.

**Comment by sykhro** [comment]
I'm afraid this can't really be merged as-is, there's way too many changes that don't really relate to the new format work and some of which that have no explanation and seem relatively dangerous

**Comment by MDHEXT** [comment]
I can't really change it. IDK visual studio code. I barely figured out how to build the software and it had an appveyor script in it.

**Comment by khang06** [comment]
oh hey, i didn't realize someone made a fork of my fork of a fork, lol. the code's like 3 years old at this point and is probably overdue for a rewrite since the original maintainer doesn't seem to be working on it anymore. i'd do it if i had enough knowledge of the rsar format

**Comment by sykhro** [comment]
Closing this for now since there's no coordination (a big refactor is also underway so this might get quite messy to rebase)
